### PR TITLE
Add DashboardLayout component to @toolpad/core

### PR DESCRIPTION
**Pull request by apedroferreira**
Isolated DashboardLayout component and its docs from https://github.com/mui/mui-toolpad/pull/3343 and implemented review feedback.

**Docs/demo:** https://deploy-preview-3554--mui-toolpad-docs.netlify.app/toolpad/core/components/dashboard-layout/

_(once we have a navigation adapter we can make the demo sidebar links work with local state to provide better examples)
(API docs generation to be added next in separate PR)_

### Summary

Please check the types in [packages/toolpad-core/src/AppProvider/AppProvider.tsx](https://github.com/mui/mui-toolpad/pull/3554/files#diff-d72d0f71caf7ef5c3f7a2250be04d58d3dc741a4e944a9c4a9f5bf1f518e6b5f) for the chosen API.

Possible to import types such as `Branding` and `Navigation` from `@toolpad/core`.
Also added `@toolpad/core/themes` where the `baseTheme` we use can be imported from, and that we use as a default theme. Eventually we can provide more themes?

### Addressed PR comments

- https://github.com/mui/mui-toolpad/pull/3343#discussion_r1597305323
- https://github.com/mui/mui-toolpad/pull/3343#discussion_r1597305679
- https://github.com/mui/mui-toolpad/pull/3343#discussion_r1597306617
- https://github.com/mui/mui-toolpad/pull/3343#discussion_r1598049998
- https://github.com/mui/mui-toolpad/pull/3343#discussion_r1598054706
- https://github.com/mui/mui-toolpad/pull/3343#discussion_r1598057358
- https://github.com/mui/mui-toolpad/pull/3343#discussion_r1598078678
- https://github.com/mui/mui-toolpad/pull/3343#discussion_r1598080006
- https://github.com/mui/mui-toolpad/pull/3343#discussion_r1598091993
- https://github.com/mui/mui-toolpad/pull/3343#discussion_r1598097992
- https://github.com/mui/mui-toolpad/pull/3343#discussion_r1598099632

### CODE REVIEW

File: docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutBasic.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import * as React from 'react';
```
Comment (Janpot) [inline docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutBasic.js:1]: Was this hand-written? I think we should port the `docs:typescript:formatted` script from core

File: docs/data/toolpad/core/api/components/dashboard-layout.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Dashboard Layout API
```
Comment (Janpot) [inline docs/data/toolpad/core/api/components/dashboard-layout.md:1]: For the API, let's port the docs generator from core?

File: docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutBasic.js (hunk 1 above)
Comment (apedroferreira) [reply docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutBasic.js:1]: I hadn't noticed that, I guess I can include it in this PR if it's simple.

File: docs/data/toolpad/core/api/components/dashboard-layout.md (hunk 2 above)
Comment (apedroferreira) [reply docs/data/toolpad/core/api/components/dashboard-layout.md:1]: We can do that in this separate PR, right? https://github.com/mui/mui-toolpad/pull/3536

File: docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutBasic.js (hunk 1 above)
Comment (apedroferreira) [reply docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutBasic.js:1]: I've added it here, seems to work fine https://github.com/mui/mui-toolpad/pull/3554/commits/9465e7462b7b45f4938a89689f1bfd14c610e7fb

File: docs/data/toolpad/core/api/components/dashboard-layout.md (hunk 2 above)
Comment (apedroferreira) [reply docs/data/toolpad/core/api/components/dashboard-layout.md:1]: Ah, the X one, I was checking the Base one, will take a look.

**Comment by apedroferreira** [comment]
This should be ready as a first version of the `DashboardLayout` - of course we can add and improve a lot in the future.
Will add the generated API docs separately, still finishing work on that but it would be easier to manage if this one is merged.

File: docs/data/toolpad/core/pages.js
Comment (Janpot) [inline docs/data/toolpad/core/pages.js:1]: I don't believe this file is expected here, see my other comment below

File: packages/toolpad-core/src/AppProvider/AppProvider.tsx [hunk 3]
```diff
@@ -1,20 +1,58 @@
...
+
+export interface Branding {
+  title?: string;
+  logo?: React.ReactNode;
+}
+
+export interface NavigationPageItem {
+  kind?: 'page';
+  title: string;
+  path?: string;
```
Comment (Janpot) [inline packages/toolpad-core/src/AppProvider/AppProvider.tsx:15]: With the nomenclature `path` I'd expect it to be the full path. Maybe we need to come up with another name e.g. like "slug"? That way we can later still add `path` if we want to add an override in the future.

```suggestion
  slug: string;
```

File: scripts/coreTypeScriptProjects.js
Comment (Janpot) [inline scripts/coreTypeScriptProjects.js:1]: 🤔  The file is already in the project, with different casing

**Review by Janpot — APPROVED** [review]

**Pull request merged**